massa-protocol-worker-2/src/messages.rs

```diff
@@ -1,11 +1,231 @@
+}
+
+impl MessagesSerializer {
+    pub fn new() -> Self {
+        Self {
+            id_serializer: U64VarIntSerializer::new(),
+            block_message_serializer: None,
+            operation_message_serializer: None,
+            endorsement_message_serializer: None,
+            peer_management_message_serializer: None,
+        }
+    }
+
+    pub fn with_block_message_serializer(
+        mut self,
+        block_message_serializer: BlockMessageSerializer,
+    ) -> Self {
+        self.block_message_serializer = Some(block_message_serializer);
+        self
+    }
+
+    pub fn with_operation_message_serializer(
+        mut self,
+        operation_message_serializer: OperationMessageSerializer,
+    ) -> Self {
+        self.operation_message_serializer = Some(operation_message_serializer);
+        self
+    }
+
+    pub fn with_endorsement_message_serializer(
+        mut self,
+        endorsement_message_serializer: EndorsementMessageSerializer,
+    ) -> Self {
+        self.endorsement_message_serializer = Some(endorsement_message_serializer);
+        self
+    }
+
+    pub fn with_peer_management_message_serializer(
+        mut self,
+        peer_management_message_serializer: PeerManagementMessageSerializer,
+    ) -> Self {
+        self.peer_management_message_serializer = Some(peer_management_message_serializer);
+        self
+    }
+}
+
+impl PeerNetMessagesSerializer<Message> for MessagesSerializer {
+    /// Serialize the id of a message
+    fn serialize_id(&self, message: &Message, buffer: &mut Vec<u8>) -> PeerNetResult<()> {
+        match message {
+            Message::BlockMessage(message) => {
+                self.id_serializer
+                    .serialize(
+                        &(MessageCategoryId::BlockHandler as u64 + message.get_id() as u64),
+                        buffer,
+                    )
+                    .map_err(|err| {
+                        PeerNetError::HandlerError.error(
+                            "MessagesSerializer",
+                            Some(format!("Failed to serialize message id: {}", err)),
+                        )
+                    })
+            },
+            Message::EndorsementMessage(message) => {
+                self.id_serializer
+                    .serialize(
+                        &(MessageCategoryId::EndorsementHandler as u64 + message.get_id() as u64),
+                        buffer,
+                    )
+                    .map_err(|err| {
+                        PeerNetError::HandlerError.error(
+                            "MessagesSerializer",
+                            Some(format!("Failed to serialize message id: {}", err)),
+                        )
+                    })
+            },
+            Message::OperationMessage(message) => {
+                self.id_serializer
+                    .serialize(
+                        &(MessageCategoryId::OperationHandler as u64 + message.get_id() as u64),
```

Comment:
I don't really like that implementation, it's very error prone.
If we add some more messages to `EndorsementMessage`, then the IDs could overlap..
A better way here would be to convert `MessageCategoryId` to `u32`, and each message type to `u32`, and create a `let id: u64 = ((msg_cat_id << 32) + msg_id) as u64;` or something of the sort.
We will never realistically get more than `2**32` messages types / categories, and it will prevent any overlap (that may be easy to provoke, and hard to debug)